_posts/2024-03-22-kraft-migration.md

```diff
@@ -0,0 +1,392 @@
+
+Removing an external system like ZooKeeper simplifies the overall architecture and removes the burden of taking care of an additional component.
+The scalability is also improved by reducing the load on the metadata store by using snapshots to avoid unbounded growth (compacted topic).
+When there is a leadership change in the quorum controller, the new leader already has all the committed metadata records so the recovery is pretty fast.
+
+Another interesting aspect of using the KRaft mode is about the role that a single node can have within the cluster itself.
+By using the new `process.roles` configuration parameter, a Kafka node can act as a `broker`, a `controller` or being in mixed-mode.
+As a `broker`, it is responsible to communicate with the Kafka clients and to store and serve data by handing read and write requests on topics.
+As a `controller`, it takes part at the quorum controller being a leader or follower, storing the cluster state and handling the metadata changes.
+Effectively, `broker` and `controller` are two different services running on the node within the JVM.
+When in mixed-mode, it gets both roles by taking care of Kafka clients communication on one hand and metadata changes on the other.
+Using the mixed-mode allows to reduce the number of nodes within the cluster compared to a ZooKeeper configuration, even if it is not recommended to be used in a production environment but only for testing or development purposes.
+
+### How to migrate from ZooKeeper to KRaft
+
+As today, it is expected to have users who are running ZooKeeper-based clusters but they are interested in migrating to KRaft mode as soon as possible in order to overcome all the limitations we have talked about.
+Furthermore, ZooKeeper is already considered as deprecated by the Apache Kafka community and its support will be removed with the 4.0 release later this year or early the next year.
+
+There is a manual procedure, made by several phases, to run in order to execute such a migration.
+The following content doesn't want to be an exhaustive description of the migration procedure but more an overview of how it is supposed to work.
+For more details, please refer to the official [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) documentation.
+
+#### Deploying the KRaft controller quorum
+
+At the beginning, we have the Kafka brokers running in ZooKeeper-mode and connected to the ZooKeeper ensemble used to store metadata.
+
+![ZooKeeper-based cluster](/assets/images/posts/2024-03-22-kraft-migration-01-zk-brokers.png)
+
+> NOTE: the green square boxed number highlights the "generation" of the nodes which are rolled more times during the process.
+
+The first step is about deploying the KRaft controller quorum which will be in charge of storing the metadata in KRaft mode.
+In general, the number of nodes will be the same as the number of the ZooKeeper nodes actually running.
```

Comment:
`the number of nodes` -> `the number of KRaft controller nodes`
`actually` Did you mean `currently`?